### PR TITLE
created objects

### PR DESCRIPTION
**Pull request by jkoren**
Created Classes for HEAT Calculator.

There are 8 classes, as described in [this comment](https://github.com/codeforboston/home-energy-analysis-tool/issues/124#issuecomment-1884053217) of [Issue 124](https://github.com/codeforboston/home-energy-analysis-tool/issues/124).

1. [Case]()
2. [Home]()
3. [Location]()
4. [Heat Load Analysis]()
5. [Natural Gas Bill]() 
6. [Natural Gas Bill Records]()
7. [Oil Propane Bill]()
8. [Oil Propane Bill Records]()

The models are stored in the /models directory
The object fields will map to database fields.

image

### CODE REVIEW

**Comment by thadk** [comment]
I used (a fork of) `pydantic2ts` to generate this Typescript of the rules engine (not their schema) in case we find it useful at any point in this process.

<details >

```typescript
/* tslint:disable */
/* eslint-disable */
/**
/* This file was automatically generated from pydantic models by running pydantic2ts.
/* Do not modify it by hand - just update the pydantic models and then re-run the script
*/

export type Dates = string[];
export type Temperatures = number[];
/**
 * Summary!G33:35
 */
export type BalancePoint = number;
/**
 * Summary!H33:35
 */
export type HeatLossRate = number;
/**
 * Summary!I33:35
 */
export type ChangeInHeatLossRate = number;
/**
 * Summary!J33:35
 */
export type PercentChangeInHeatLossRate = number;
/**
 * Summary!K33:35
 */
export type StandardDeviation = number;
export type Records = BalancePointGraphRow[];
/**
 * DHW!B4
 */
export type NumberOfOccupants = number;
/**
 * DHW!B5
 */
export type EstimatedWaterHeatingEfficiency = number;
/**
 * DHW!B6
 */
export type StandByLosses = number;
/**
 * Natural Gas!A
 */
export type PeriodStartDate = string;
/**
 * Natural Gas!B
 */
export type PeriodEndDate = string;
/**
 * Natural Gas!D
 */
export type UsageTherms = number;
export type AnalysisType = 0 | 1 | -1;
export type Records1 = NaturalGasBillingRecordInput[];
export type PeriodStartDate1 = string;
export type PeriodEndDate1 = string;
export type Usage = number;
/**
 * Oil-Propane!B
 */
export type PeriodEndDate2 = string;
/**
 * Oil-Propane!C
 */
export type Gallons = number;
/**
 * Oil-Propane!F
 */
export type InclusionOverride = boolean | null;
export type Records2 = OilPropaneBillingRecordInput[];
/**
 * Oil-Propane!B6
 */
export type PrecedingDeliveryDate = string;
/**
 * Summary!B10
 */
export type LivingArea = number;
/**
 * Enum for fuel types. Values are BTU per usage
 */
export type FuelType = 100000 | 139600 | 91333;
/**
 * Summary!B12
 */
export type HeatingSystemEfficiency = number;
/**
 * Summary!B17
 */
export type ThermostatSetPoint = number;
/**
 * Summary!B18
 */
export type SetbackTemperature = number | null;
/**
 * Summary!B19
 */
export type SetbackHoursPerDay = number | null;
/**
 * TDesign
 */
export type DesignTemperature = number;
/**
 * Summary!B20
 */
export type EstimatedBalancePoint = number;
/**
 * Summary!B15
 */
export type OtherFuelUsage = number;
/**
 * Summary!B24
 */
export type AverageIndoorTemperature = number;
/**
 * Summary!B25
 */
export type DifferenceBetweenTiAndTbp = number;
/**
 * Summary!B26
 */
export type DesignTemperature1 = number;
/**
 * Summary!B27
 */
export type WholeHomeHeatLossRate = number;
/**
 * Summary!B28
 */
export type StandardDeviationOfHeatLossRate = number;
/**
 * Summary!B29
 */
export type AverageHeatLoad = number;
/**
 * Summary!B30
 */
export type MaximumHeatLoad = number;
export type Dates1 = string[];
export type Temperatures1 = number[];

/**
 * From Summary page
 */
export interface BalancePointGraph {
  records: Records;
}
/**
 * From Summary page
 */
export interface BalancePointGraphRow {
  balance_point: BalancePoint;
  heat_loss_rate: HeatLossRate;
  change_in_heat_loss_rate: ChangeInHeatLossRate;
  percent_change_in_heat_loss_rate: PercentChangeInHeatLossRate;
  standard_deviation: StandardDeviation;
}
/**
 * From DHW (Domestic Hot Water) Tab
 */
export interface DhwInput {
  number_of_occupants: NumberOfOccupants;
  estimated_water_heating_efficiency: EstimatedWaterHeatingEfficiency;
  stand_by_losses: StandByLosses;
}
/**
 * From Natural Gas tab. Container for holding all rows of the billing input table.
 */
export interface NaturalGasBillingInput {
  records: Records1;
}
/**
 * From Natural Gas tab. A single row of the Billing input table.
 */
export interface NaturalGasBillingRecordInput {
  period_start_date: PeriodStartDate;
  period_end_date: PeriodEndDate;
  usage_therms: UsageTherms;
  /**
   * Natural Gas!E
   */
  inclusion_override: AnalysisType | null;
}
export interface NormalizedBillingPeriodRecordInput {
  period_start_date: PeriodStartDate1;
  period_end_date: PeriodEndDate1;
  usage: Usage;
  inclusion_override: AnalysisType | null;
}
/**
 * From Oil-Propane tab. Container for holding all rows of the billing input table.
 */
export interface OilPropaneBillingInput {
  records: Records2;
  preceding_delivery_date: PrecedingDeliveryDate;
}
/**
 * From Oil-Propane tab
 */
export interface OilPropaneBillingRecordInput {
  period_end_date: PeriodEndDate2;
  gallons: Gallons;
  inclusion_override: InclusionOverride;
}
/**
 * From Summary Tab
 */
export interface SummaryInput {
  living_area: LivingArea;
  /**
   * Summary!B11
   */
  fuel_type: FuelType;
  heating_system_efficiency: HeatingSystemEfficiency;
  thermostat_set_point: ThermostatSetPoint;
  setback_temperature: SetbackTemperature;
  setback_hours_per_day: SetbackHoursPerDay;
  design_temperature: DesignTemperature;
}
/**
 * From Summary tab
 */
export interface SummaryOutput {
  estimated_balance_point: EstimatedBalancePoint;
  other_fuel_usage: OtherFuelUsage;
  average_indoor_temperature: AverageIndoorTemperature;
  difference_between_ti_and_tbp: DifferenceBetweenTiAndTbp;
  design_temperature: DesignTemperature1;
  whole_home_heat_loss_rate: WholeHomeHeatLossRate;
  standard_deviation_of_heat_loss_rate: StandardDeviationOfHeatLossRate;
  average_heat_load: AverageHeatLoad;
  maximum_heat_load: MaximumHeatLoad;
}
export interface TemperatureInput {
  dates: Dates1;
  temperatures: Temperatures1;
}

```

<details >

**Comment by thadk** [comment]
Have you used types/interfaces instead of classes before? I wanted to talk to you about that but we ran out of time.
Essentially I’m trying to realize what is the benefit for us in stateful class instance objects as here vs simple stateless plain typescript objects re-created much like the classes in your PRs?

**Comment by thadk** [comment]
In Slack, Jeff asked the difference between `type` syntax, `interface` syntax and even `class` syntax. I found an awesome answer that explains the syntax and other (sorta minor) differences around choosing Types vs Interfaces https://stackoverflow.com/a/52682220

**Comment by jkoren** [comment]
I read the article, and updated PR 133 to use interfaces.  good suggestion!

File: heat-stack/model/Cases.tsx
Comment (alanisaac) [inline heat-stack/model/Cases.tsx:1]: `nitpick`: I notice that there are some inconsistencies in the naming of these interface files with regards to singular vs plural.  e.g. this is `Cases`, but `Location` is not `Locations`.  Should that be consistent across all these files?

File: heat-stack/model/HeatLoadAnalysis.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export interface HeatLoadAnalysis {
+	rules_engine_version: string
+	estimatedBalancePoint: number
+	otherFuelUsage: number
+	averageIndoorTemperature: number
+	differenceBetweenBAndTBP: number
```
Comment (alanisaac) [inline heat-stack/model/HeatLoadAnalysis.tsx:6]: `suggestion`: I think the first `BA` acronym here isn't quite right and should be `differenceBetweenTiAndTbp`

File: heat-stack/model/Homes.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export interface Home {
+	livingArea: number
+	fuelType: string // number?
```
Comment (alanisaac) [inline heat-stack/model/Homes.tsx:3]: `suggestion`: the database won't have a concept of it, but I might use an [enum ](https://www.typescriptlang.org/docs/handbook/enums.html) here.

File: heat-stack/model/NaturalGasBillRecords.tsx [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface NaturalGasBillRecords {
+	provider: number
+	periodStartDate: Date
+	periodEndDate: Date
+	usageTherms: number
+	inclusionOverride: number
```
Comment (alanisaac) [inline heat-stack/model/NaturalGasBillRecords.tsx:6]: `suggestion`: the database won't have a concept of it, but I might use an [enum ](https://www.typescriptlang.org/docs/handbook/enums.html) here.

File: heat-stack/model/OilPropaneBillRecords.tsx [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export interface OilPropaneBillRecords {
+	provider: number
+	periodStartDate: Date
+	periodEndDate: Date
+	gallons: number
+	inclusionOverride: number
```
Comment (alanisaac) [inline heat-stack/model/OilPropaneBillRecords.tsx:6]: `suggestion`: the database won't have a concept of it, but I might use an [enum ](https://www.typescriptlang.org/docs/handbook/enums.html) here.

**Review by alanisaac — COMMENTED** [review]
A general question: to what extent are we loading objects from the database in isolation vs together as an aggregate?  e.g. there are no references between objects in the current models.  I might expect, for example, that `NaturalGasBill` is a aggregate root, and might have an array of `NaturalGasBillRecords` as part of the object.

File: heat-stack/model/HeatLoadAnalysis.tsx [hunk 5]
```diff
@@ -0,0 +1,12 @@
+export interface HeatLoadAnalysis {
+	rules_engine_version: string
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/HeatLoadAnalysis.tsx:2]: good catch

File: heat-stack/model/HeatLoadAnalysis.tsx (hunk 1 above)
Comment (jkoren) [reply heat-stack/model/HeatLoadAnalysis.tsx:6]: ok

File: heat-stack/model/Homes.tsx (hunk 2 above)
Comment (jkoren) [reply heat-stack/model/Homes.tsx:3]: I think we can do 
`fuelType: 'Natural Gas' | 'Oil'`

File: heat-stack/model/OilPropaneBillRecords.tsx [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export interface OilPropaneBillRecords {
+	provider: number
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/OilPropaneBillRecords.tsx:2]: makes sense

File: heat-stack/model/OilPropaneBill.tsx [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export interface OilPropaneBill {
+	provider: number
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/OilPropaneBill.tsx:2]: changed from "number" to "string"

File: heat-stack/model/Homes.tsx [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+	designTemperatureOverride: number
+	heatingSystemEfficiency: number
+	thermostatSetPoint: number
+	setbackTemperature: number
+	setBackHoursPerDay: number
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/Homes.tsx:8]: updated

File: heat-stack/model/HeatLoadAnalysis.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/HeatLoadAnalysis.tsx:2]: good catch - done

File: heat-stack/model/OilPropaneBill.tsx [hunk 9]
```diff
@@ -0,0 +1,4 @@
...
+	precedingDeliverDate: Date
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/OilPropaneBill.tsx:3]: updated

File: heat-stack/model/NaturalGasBillRecords.tsx [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export interface NaturalGasBillRecords {
+	provider: number
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/NaturalGasBillRecords.tsx:2]: removed

File: heat-stack/model/NaturalGasBillRecords.tsx (hunk 3 above)
Comment (jkoren) [reply heat-stack/model/NaturalGasBillRecords.tsx:6]: what should the options be?

File: heat-stack/model/OilPropaneBillRecords.tsx (hunk 4 above)
Comment (jkoren) [reply heat-stack/model/OilPropaneBillRecords.tsx:6]: @alanisaac what should the options be?

File: heat-stack/model/Homes.tsx (hunk 2 above)
Comment (jkoren) [reply heat-stack/model/Homes.tsx]: @alanisaac should there be other options?

File: heat-stack/model/Cases.tsx
Comment (jkoren) [reply heat-stack/model/Cases.tsx:1]: Good catch.  I made all the models singular.

File: heat-stack/model/NaturalGasBillRecords.tsx (hunk 3 above)
Comment (alanisaac) [reply heat-stack/model/NaturalGasBillRecords.tsx]: The rules engine models it like so: https://github.com/codeforboston/home-energy-analysis-tool/blob/main/rules-engine/src/rules_engine/pydantic_models.py#L13-L16

```py
class AnalysisType(Enum):
    DO_NOT_INCLUDE = 0
    INCLUDE = 1
    INCLUDE_IN_OTHER_ANALYSIS = -1
```

I'd use those names for consistency (in whatever the right casing is for Javascript).

File: heat-stack/model/OilPropaneBillRecords.tsx (hunk 4 above)
Comment (alanisaac) [reply heat-stack/model/OilPropaneBillRecords.tsx]: See https://github.com/codeforboston/home-energy-analysis-tool/pull/133#discussion_r1454343956

File: heat-stack/model/Homes.tsx (hunk 2 above)
Comment (alanisaac) [reply heat-stack/model/Homes.tsx]: The options in the spreadsheet are:

* Natural Gas
* Fuel Oil
* Propane

There is a calculation distinction between fuel oil and propane.

Comment (jkoren) [reply heat-stack/model/Homes.tsx]: updated to `fuelType: 'Natural Gas' | 'Oil' | 'Propane'`

File: heat-stack/model/NaturalGasBillRecords.tsx (hunk 3 above)
Comment (jkoren) [reply heat-stack/model/NaturalGasBillRecords.tsx]: updated to `inclusionOverride: 'Include' | 'Do not include' | 'Include in other analysis'`

File: heat-stack/model/OilPropaneBillRecords.tsx (hunk 4 above)
Comment (jkoren) [reply heat-stack/model/OilPropaneBillRecords.tsx]: updated to `inclusionOverride: 'Include' | 'Do not include' | 'Include in other analysis'`

File: heat-stack/model/Location.tsx [hunk 11]
```diff
@@ -0,0 +1,8 @@
+export interface Location {
+	address: string
+	address_line_2: string
```
[the remark this answers is not in the document]
Comment (jkoren) [reply heat-stack/model/Location.tsx:3]: done

**Comment by jkoren** [comment]
@alanisaac Yes NaturalGasBill is an aggregate root and would have an array of NaturalGasBillRecords as part of the object.

**Comment by thadk** [comment]
@jkoren What if we combine and move all the typescript files from `heat-stack/model` in to the `heat-stack/types/index.d.ts` or `heat-stack/types/index.ts` file if that doesn't work for some reason?

**Comment by thadk** [comment]
@jkoren I think we should create `heat-stack/types/index.d.ts` with the combined set of those. I think combining related types is pretty common. 

Also I think `heat-stack/models` is a bit inaccurate naming since the database schemas won't be controlled by these files, only the types. 

The SQLite database schema is controlled by the `schema.prisma` file I believe.

**Comment by jkoren** [comment]
@thadk combined interfaces into single file

**Comment by thadk** [comment]
@jkoren Oop, tests aren't passing though. Looks like some references might still be on the old setup.

**Comment by thadk** [comment]
Thanks so much Jeff and all on this one. 
- We went over this with the rules engine team yesterday and they used a version of it to do a complete test with Epic Stack forms and Pyodide and the Rules Engine. We're going to switch to this version, maybe with some small tweaks to match with our latest thinking on the call format to the Rules Engine.

**Pull request merged**